tests/integration/DistributedCacheInvalidationTest.php

```diff
@@ -0,0 +1,280 @@
+        // Set a cache version in Redis
+        $redis->connection('fof.cache')->set('flarum:cache:version', time());
+
+        // Directly invoke the middleware logic via reflection to simulate version detection
+        // Note: In real PHP-FPM workers, the middleware detects version changes across requests
+        // In tests with process isolation, we test the invalidation logic directly
+        $reflectionClass = new \ReflectionClass($middleware);
+
+        $shouldInvalidateMethod = $reflectionClass->getMethod('shouldInvalidateLocal');
+        $shouldInvalidateMethod->setAccessible(true);
+
+        $invalidateMethod = $reflectionClass->getMethod('invalidateLocalCaches');
+        $invalidateMethod->setAccessible(true);
+
+        // Call invalidation directly
+        $invalidateMethod->invoke($middleware);
+
+        // Files should be cleared
+        $this->assertFileDoesNotExist($formatterCache.'/test.php', 'test.php should be deleted');
+        $this->assertFileDoesNotExist($localeCache.'/en.php', 'en.php should be deleted');
+        $this->assertFileDoesNotExist($viewsCache.'/test.blade.php', 'test.blade.php should be deleted');
+    }
+
+    /**
+     * @test
+     */
+    public function middleware_throttles_redis_checks()
+    {
+        $this->requiresRedis();
+
+        $redis = $this->app()->getContainer()->make(Factory::class);
+
+        // Set initial version
+        $redis->connection('fof.cache')->set('flarum:cache:version', time());
+
+        // Make multiple requests quickly
+        $response1 = $this->send($this->request('GET', '/'));
+        $response2 = $this->send($this->request('GET', '/'));
+        $response3 = $this->send($this->request('GET', '/'));
+
+        // All should succeed
+        $this->assertEquals(200, $response1->getStatusCode());
+        $this->assertEquals(200, $response2->getStatusCode());
+        $this->assertEquals(200, $response3->getStatusCode());
+
+        // The middleware should have only checked Redis once due to throttling
+        // (We can't directly verify this without instrumentation, but we verify it doesn't cause errors)
+    }
+
+    /**
+     * @test
+     */
+    public function middleware_handles_missing_redis_gracefully()
+    {
+        // Don't require Redis for this test - we want to test the error handling
+
+        // Create a version of the app without Redis
+        $this->config('redis.connection', null);
+
+        // Request should still work even if Redis is unavailable
+        $response = $this->send($this->request('GET', '/'));
+        $this->assertEquals(200, $response->getStatusCode());
+    }
+
+    /**
+     * @test
+     */
+    public function cache_version_increases_on_each_clear()
+    {
+        $this->requiresRedis();
+
+        $redis = $this->app()->getContainer()->make(Factory::class);
+        $events = $this->app()->getContainer()->make(Dispatcher::class);
+
+        // First clear
+        $events->dispatch(new ClearingCache());
```

Comment:
Is this blocking? Otherwise, we should probably wait until the cache clear was sucessfull as it may take longer than the 1-second sleep below.